crates/swc_plugin_runner/src/lib.rs

```diff
@@ -1,41 +1,87 @@
-use abi_stable::{
-    library::RootModule,
-    std_types::{RResult, RStr},
-};
-use anyhow::{anyhow, Context, Error};
-use rplugin::StableAst;
+use anyhow::{Context, Error};
+use resolve::PluginCache;
 use std::path::Path;
 use swc_ecma_ast::Program;
-use swc_plugin_js_api::SwcJsPluginRef;
+use wasmer::{imports, Instance, Module, Store};
+use wasmer_cache::{Cache, Hash};
 
 pub mod resolve;
 
-pub fn apply_js_plugin(
-    plugin_name: &str,
-    path: &Path,
-    config_json: &str,
-    program: Program,
-) -> Result<Program, Error> {
-    (|| -> Result<_, Error> {
-        let plugin_rt = swc_common::plugin::get_runtime_for_plugin(plugin_name.to_string());
+/// Load plugin from specified path.
+/// If cache is provided, it'll try to load from cache first to avoid
+/// compilation.
+fn load_plugin(plugin_path: &Path, cache: &mut Option<PluginCache>) -> Result<Instance, Error> {
+    let wasmer_store = Store::default();
```

Comment:
I think it's better to cache file content.

`SwcJsPluginRef` did this internally, but now we are reading the `.wasm` file every time `apply_js_plugin` is called.

Reply:
How does cache invalidation works if we store full file contents and do not read again? Having hash from full file is in my opinion most correct way to ensure cache can be invalidated when file changes, like having new versions. Other way to calculate hash (name + path + mtime...) is quite error prone afaik.

There is def. overhead around reading file over and over again, but I expect it's somewhat acceptable for the first implementation and can optimize later if we heavily offended by those.

Reply:
(fyi, wasmer itself does same for the hashing : https://github.com/wasmerio/wasmer/blob/d5a444d61f74babc438cf669470ca7e62d891a46/lib/cli/src/commands/run.rs#L225-L246)

Reply:
> we store full file contents and do not read again

This was true for previous version, and I'm not sure if it's useful to support updating plugins within the same process.
Hmm... Might be useful if we have a server mode.

Reply:
No, current cache targets filesystem so this is permanant. So it becomes matter to get right hash to load up compiled binary for the subsequent process invocation.

For example, if we incorrectly calculate hash like `name + path` which'll be same over different version, user'll see unexpected behavior by we serve old versions of cache. mtime is being used for those purpose for some cases, but that's also error prone on the CI systems.

Reply:
What do you mean by `current`?
This PR or previous one?

Reply:
I meant this PR.

Pseudo workflow like

1. swc runs plugin A for the first time, cache bytecode (and exit)
2. next time swc runs, swc finds hash for A
2.1 cache hits - returns compiled bytecode from filesystem cache
2.2 cache miss - consider this is same as 1, go back to 1

The part in discussion is incorrect 2.1 
- plugin A.1 (new version of A) hash calculated to plugin A, swc returns compiled bytecode of A instead.

Reply:
Also, we need to cache `dual` if we want to store file contents as well - first we need to store raw file contents, secondly need to store compiled bytecode which is actual cached contents we need to serve to avoid compilation.

Reply:
Ah, I was talking about `std::fs::read(plugin_path)` being called multiple times, within a single invocation.

Reply:
ugh I misunderstood your comment completely then 😅 

But also - how frequent cases will there be if the same plugin will be specified more than once in a config? Will we allow it even? Unless I misunderstand again, reading file should occur for any unique plugin binaries.

Reply:
The problem is not a single plugin being specified multiple times. If we compile 3000 files, this will read the wasm file at least 3000 times.

Reply:
oh, right 🤦 I somehow thought this is being called per each loading phase, not for the each transform.